### PR TITLE
Start adding store-and-forward direct-to-device messaging

### PR DESCRIPTION
**Pull request by NegativeMjark**
Implement the local part of direct-to-device messaging intended for exchanging end-to-end keys and other device-to-device signalling.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/sendtodevice.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+from synapse.http import servlet
+from synapse.rest.client.v1.transactions import HttpTransactionStore
+from ._base import client_v2_patterns
+
+logger = logging.getLogger(__name__)
+
+
+class SendToDeviceRestServlet(servlet.RestServlet):
+    PATTERNS = client_v2_patterns(
+        "/sendToDevice/(?P<message_type>[^/]*)/(?P<txn_id>[^/]*)$",
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/sendtodevice.py]: We use underscores fore APIs?

File: synapse/storage/deviceinbox.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                return []
+            sql = (
+                "SELECT user_id, device_id FROM devices"
+                " WHERE user_id = ? AND device_id IN ("
+                + ",".join("?" * len(devices))
+                + ")"
+            )
+            # TODO: Maybe this needs to be done in batches if there are
+            # too many local devices for a given user.
+            args = [user_id] + devices
```
Comment (erikjohnston) [inline synapse/storage/deviceinbox.py]: there is a `_simple_select_many_batch`

**Comment by erikjohnston** [comment]
LGTM. Tests look like you need to merge in from develop.

File: synapse/rest/client/v2_alpha/sendtodevice.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/sendtodevice.py]: camelcased:
```
/_matrix/client/r0/createRoom
/_matrix/client/r0/initialSync
/_matrix/client/r0/publicRooms
/_matrix/client/r0/voip/turnServer
^/_matrix/client/r0/rooms/[^/]*/initialSync$
```

underscores:
```
^/_matrix/client/r0/profile/[^/]*/avatar_url$
^/_matrix/client/r0/user/[^/]*/account_data/[^/]*$
```

lowercase:
```
^/_matrix/client/r0/pushrules/
^/_matrix/client/r0/profile/[^/]*/displayname$
```

Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/sendtodevice.py]: I retract my previous statement, and replace it with: your style isn't inconsistent enough.

**Pull request merged**